src/lib/HST/Feature/GuardElimination.hs

```diff
@@ -1,255 +1,269 @@
   S.NoSrcSpan
-  [S.Match S.NoSrcSpan name [] (S.UnGuardedRhs S.NoSrcSpan e) Nothing]
-toDecl _ _ = error "GuardElimination.toDecl: Variable pattern expected"
+  [S.Match S.NoSrcSpan name [] (S.UnGuardedRhs S.NoSrcSpan expr) Nothing]
 
--- Folds the list of GExps to declarations.
-foldGEqs
-  :: Member Fresh r
-  => [S.Pat a]                   -- fresh variables for the case exps
-  -> ([S.Decl a], S.Pat a)       -- startcase ([], first generated Pattern)
-  -> [GExp a]                    -- list of pattern + rhs pair
-  -> Sem r ([S.Decl a], S.Pat a) -- a list of declarations for the let binding
-                                          -- and a free Variable for the error case
-foldGEqs vs = foldM (createDecl vs)
-
--- Generates a varbinding and a new variable for the next var binding
-createDecl
-  :: Member Fresh r
-  => [S.Pat a]                   -- generated variables
-  -> ([S.Decl a], S.Pat a)       -- (current decls , variable for let binding)
-  -> GExp a                      -- pairs of pattern to match against and a guarded Rhs
-  -> Sem r ([S.Decl a], S.Pat a) -- var bindings , variable for next match
-createDecl vs (decl, p) (ps, rhs) = do
-  nVar <- freshVarPat genericFreshPrefix
-  let varExp = CA.translatePVar nVar
-  iexp <- rhsToIf rhs varExp
-  let cexp  = createCase iexp varExp (zip vs ps)
-  let ndecl = toDecl p cexp
-  return (ndecl : decl, nVar)
-
--- TODO refactor to higher order
--- Generates a recursive case expression for each variable and pattern pair
-createCase
-  :: S.Exp a              -- ifThenElse
-  -> S.Exp a              -- the other pattern (in case pattern match or guard fails)
-  -> [(S.Pat a, S.Pat a)] -- Patterns to match (PVar , Pattern)
+-------------------------------------------------------------------------------
+-- @case@ Expressions                                                        --
+-------------------------------------------------------------------------------
+
+-- | Generates nested case expression for each variable and pattern pair.
+--
+--   @'generateNestedCases' e f [(x₁, p₁), …, (xₙ, pₙ)]@ produces an expression
+--   of the following form.
+--
+--   > case x₁ of { p₁ -> (… case xₙ of { pₙ -> e ; _ -> f }  …) ; _ -> f }
+generateNestedCases
+  :: S.Exp a              -- ^ Expression to use if all pattern match.
+  -> S.Exp a              -- ^ Expression to use if any pattern does not match.
+  -> [(S.Exp a, S.Pat a)] -- ^ Expression/pattern pairs to match.
   -> S.Exp a
--- createCase i next vps
---   = foldr (\(v,p) next ->
---       Case () (translatePVar v)
---               [S.alt p res, S.alt B.wildcard next]) i vps
-createCase i _    []             = i
-createCase i next ((v, p) : vps) = S.Case
-  S.NoSrcSpan
-  (CA.translatePVar v)
-  [S.alt p (createCase i next vps), S.alt (S.PWildCard S.NoSrcSpan) next]
+generateNestedCases successExpr failExpr = foldr generateNestedCase successExpr
+ where
+  {- generateNestedCase :: (S.Exp a, S.Pat a) -> S.Exp a -> S.Exp a -}
```

Comment:
Why is the type signature commented out?

Reply:
Since `failExpr` is used in `generateNestedCase` the `a` in the type signature of `generateNestedCase` must be the same `a` as in the type signature of `generateNestedCases`. Due to Haskells type variable scoping rules, it is not possible to specify the type of `geneateNestedCase` without language extensions. Thus, the type signature must be commented out as suggested by our [Haskell Styleguide](https://github.com/FreeProving/guidelines/blob/main/CONTRIBUTING.md#add-type-signatures-for-all-function-declarations).